### PR TITLE
Add a static single-producer single-consumer queue

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

This is needed for efficient intra-node data transfers when using a nodegroup DG element setup.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Parallel/StaticSpscQueue.hpp [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+  /// is at capacity and does not push the new object, otherwise returns `true`.
+  ///
+  /// Uses `try_emplace()` internally.
+  template <typename P, Requires<std::is_constructible_v<T, P&&>> = nullptr>
+  [[nodiscard]] bool try_push(P&& v) noexcept(
+      std::is_nothrow_constructible_v<T, P&&>) {
+    return try_emplace(std::forward<P>(v));
+  }
+
+  /// Removes the first element from the queue and returns it.
```
Comment (wthrowe) [inline src/Parallel/StaticSpscQueue.hpp]: This just returns it, doesn't remove.

**Review by wthrowe — CHANGES_REQUESTED** [review]
The internal caching logic all looks right.

File: src/Parallel/StaticSpscQueue.hpp [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+  /// Returns the capacity of the queue.
+  [[nodiscard]] size_t capacity() const noexcept { return capacity_ - 1; }
+
+ private:
+  static constexpr size_t capacity_ = Capacity + 1;
+  std::array<T, capacity_ + 2 * padding_> data_{};
+
+  // Align to cache line size in order to avoid false sharing
+  // readIndexCache_ and writeIndexCache_ is used to reduce the amount of cache
```
Comment (kidder) [inline src/Parallel/StaticSpscQueue.hpp]: update these to correct names

File: src/Parallel/StaticSpscQueue.hpp [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+  // coherency traffic
+  alignas(cache_line_size_) std::atomic<size_t> write_index_{0};
+  alignas(cache_line_size_) size_t read_index_cache_{0};
+  alignas(cache_line_size_) std::atomic<size_t> read_index_{0};
+  alignas(cache_line_size_) size_t write_index_cache_{0};
+
+  // Padding to avoid adjacent allocations from sharing a cache line with
+  // writeIndexCache_
```
Comment (kidder) [inline src/Parallel/StaticSpscQueue.hpp]: update name

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
Thanks for the reviews! I rebased and just squashed the minor documentation changes.

**Review by wthrowe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**